design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,264 @@
+
+Currently, this is handled by defining multiple variables in Conjur. Each variable contains a single value, such as a username, a password, etc. To bundle these pieces together, we use a naming convention. For example:
+
+```yaml
+- !policy
+  id: my-policy
+  body:
+    - &vars
+      - !variable mysql-db-creds/password
+      - !variable mysql-db-creds/username
+      - !variable mysql-db-creds/address
+      - !variable mysql-db-creds/port
+
+    - !host my-app
+    
+    - !permit
+      role: !host my-app
+      privileges: [ read,execute ]
+      resource: *vars
+```
+
+## Proposed Solution
+
+The proposed solution is to add the ability to use JSON schemas and link them to variables, for which we want validation of their content. This will allow a single variable to store multiple, related values. While it's already possible today, to store JSON structured information in a variable, this proposed feature adds two added values:
+
+- The JSON structure is validated and enforced. If the input is invalid, the user will get a proper error message and the variable update will be denied.
+- The JSON structure will not be encrypted completely, but only the sensitive information. This will allow future seach capabilities based on the non sensitive information.
+
+### User Experience
+
+Let's have a look at the following example:
+
+```yaml
+- !policy
+  id: my-policy
+  body:
+  - !variable
+    id: mysql-db-creds
+    schema: !variable /conjur/schemas/mysql-schema
+
+  - !host my-app
+  
+  - !permit
+    role: !host my-app
+    privileges: [ read,execute ]
+    resource: !variable mysql-db-creds
+```
+
+And
+
+```yaml
+- !policy
+  id: conjur/schemas
+  body:
+  - &schemas
+    - !variable mysql-schema
+  
+  - !permit
+    role: !group /conjur/all
+    privileges: [ read,execute ]
+    resource: !variable *schemas
+```
+
+In the example above, the `/conjur/schemas/mysql-schema` variable will be updated with the following value:
+
+```json
+{
+  "$id": "https://cyberark.com/mysql.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "MySQL Creds",
+  "type": "object",
+  "properties": {
+    "username": {
+      "type": "string",
+      "description": "The MySQL username"
+    },
+    "password": {
+      "type": "string",
+      "description": "The MySQL password",
+      "minLength": 12,
+      "maxLength": 40,
+      "pattern": "[A-Za-z0-9]"
+    },
+    "address": {
+      "type": "string",
+      "description": "The MySQL address"
+    },
+    "port": {
+      "description": "The MySQL port number",
+      "type": "integer",
+      "minimum": 1,
+      "maximum": 65535
+    }
+  },
+  "required": ["username", "password"],
+  "additionalProperties": false,
+  "secrets": [
+      "password",
+    ]
+}
+```
+
+And the `mysql-db-creds` variable could be updated with the following value:
```

Comment:
Something that I thought about here, is how easy is it for a user to change their schema? What will happen if they change the schema after variables are loaded. Will they get a list of current variables that do not abide by the newly generated/changed schema?

Can a user have multiple schema's attached to their variables? Something like
```
- !policy
  id: my-policy
  body:
  - !variable
    id: mysql-db-creds
    schema: !variable /conjur/schemas/mysql-schema
    schema:  !variable /conjur/schemas/another-mysql-schema
```

Reply:
Great questions, I'll clarify better in the design. My thoughts are:

1. No, you can't use more than one schema. I think it will over complicate our functionality, and will not provide a worthwhile value.
1. The schema is validated before the secret value is updated. I think that if the schema is changing, it will not affect the existing variables, but will only affect future updates.